mls-rs/src/key_package/builder.rs

```diff
@@ -0,0 +1,309 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// Copyright by contributors to this project.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+use core::fmt::Debug;
+
+use alloc::vec;
+use mls_rs_codec::{MlsDecode, MlsEncode, MlsSize};
+use mls_rs_core::error::IntoAnyError;
+use mls_rs_core::extension::MlsExtension;
+use mls_rs_core::key_package::KeyPackageData;
+
+use crate::client::MlsError;
+use crate::client_config::ClientConfig;
+use crate::Client;
+use crate::{
+    crypto::{HpkeSecretKey, SignatureSecretKey},
+    group::framing::MlsMessagePayload,
+    identity::SigningIdentity,
+    protocol_version::ProtocolVersion,
+    signer::Signable,
+    tree_kem::{
+        leaf_node::{ConfigProperties, LeafNode},
+        Capabilities, Lifetime,
+    },
+    CipherSuiteProvider, ExtensionList, MlsMessage,
+};
+
+use super::{KeyPackage, KeyPackageRef};
+
+#[derive(Clone, Debug)]
+pub struct KeyPackageBuilder<'a, CP> {
+    protocol_version: ProtocolVersion,
+    cipher_suite_provider: CP,
+    signing_identity: Option<SigningIdentity>,
+    signing_key: Option<&'a SignatureSecretKey>,
+    key_package_extensions: ExtensionList,
+    leaf_node_extensions: ExtensionList,
+    validity_sec: u64,
+    // This I feel like can still be fixed for client as it rarely changes?
+    capabilities: Capabilities,
+}
+
+impl<'a, CP> KeyPackageBuilder<'a, CP> {
+    pub fn signing_data(
+        self,
+        signing_identity: SigningIdentity,
+        signing_key: &'a SignatureSecretKey,
+    ) -> Self {
+        Self {
+            signing_identity: Some(signing_identity),
+            signing_key: Some(signing_key),
+            ..self
+        }
+    }
+
+    pub fn with_key_package_extension<T: MlsExtension>(
+        mut self,
+        extension: T,
+    ) -> Result<Self, MlsError> {
+        self.key_package_extensions.set_from(extension)?;
+
+        Ok(self)
+    }
+
+    pub fn with_leaf_node_extension<T: MlsExtension>(
+        mut self,
+        extension: T,
+    ) -> Result<Self, MlsError> {
+        self.leaf_node_extensions.set_from(extension)?;
+
+        Ok(self)
+    }
+
+    pub fn valid_for_sec(self, validity_sec: u64) -> Self {
+        Self {
+            validity_sec,
+            ..self
+        }
+    }
+}
+
+impl<CP: CipherSuiteProvider> KeyPackageBuilder<'_, CP> {
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub async fn build(self) -> Result<KeyPackageGeneration, MlsError> {
+        let (signing_identity, signing_key) = self
+            .signing_identity
+            .zip(self.signing_key)
+            .ok_or(MlsError::SignerNotFound)?;
```

Comment:
I feel like this is an anti-pattern for builder design pattern. If the signer is required, then it should be set in the constructor

Reply:
Moved to constructor.